File: optimizers/Adafactor.py
from .Optimizer import Optimizer, np


class Adafactor(Optimizer):
    """
    Adafactor — адаптивный метод с сублинейной памятью; для параметра-вектора
    используется нефакторизованная оценка второго момента.

    Первоисточник: Shazeer N., Stern M. "Adafactor: Adaptive Learning Rates
    with Sublinear Memory Cost". arXiv:1804.04235, 2018.
    https://arxiv.org/abs/1804.04235
    Реализация следует алгоритму из документации PyTorch:
    https://docs.pytorch.org/docs/stable/generated/torch.optim.Adafactor.html
    """

    step: float
    variance: np.ndarray

    def __init__(
        self,
        initial_x: np.ndarray,
        function,
        lr: float = 0.1,
        beta2_decay: float = -0.8,
        eps1: float = 1e-15,
        eps2: float = 0.001,
        d: float = 1.0,
        weight_decay: float = 0.0,
    ) -> None:
        params = dict(lr=lr, beta2_decay=beta2_decay, eps1=eps1, eps2=eps2, d=d, weight_decay=weight_decay)
        super().__init__(initial_x, function, params)

    def next_point(self) -> tuple[np.ndarray, float]:
        gradient = self.function.grad(self.x)
        self.step += 1
        # для одномерного параметра PyTorch использует нефакторизованную оценку второго момента
        one_minus_beta2_t = self.step ** self.params["beta2_decay"]
        rho_t = min(self.params["lr"], 1 / self.step**0.5)
        alpha = max(self.params["eps2"], float(np.linalg.norm(self.x)) / self.x.size**0.5) * rho_t
        x = self.x * (1 - self.params["lr"] * self.params["weight_decay"])
        self.variance = self.variance + (gradient**2 - self.variance) * one_minus_beta2_t
        update = gradient / np.sqrt(np.maximum(self.variance, self.params["eps1"] ** 2))
        denom = max(1.0, float(np.linalg.norm(update)) / (update.size**0.5 * self.params["d"]))
        next_x = x - alpha / denom * update
        return self.move_next(next_x)

    def reset(self) -> None:
        super().reset()
        self.step = 0.0
        self.variance = np.zeros([2])

File: src/shared/lib/optimization-engine/optimizers/adafactor.ts
import type { OptimizerDescriptor } from "./types";
import { norm2, zero2, type Vec2 } from "@shared/lib/optimization-engine/linalg";
import { gradient } from "@shared/lib/optimization-engine/functions";

const DIM_SQRT = Math.sqrt(2);

export const adafactorOptimizer: OptimizerDescriptor = {
  name: "Adafactor",
  params: {
    lr: { default: 0.1, description: "верхняя граница относительного шага rho_t = min(lr, 1/sqrt(t))" },
    beta2Decay: {
      default: -0.8,
      description: "показатель затухания второго момента: вес свежего квадрата градиента равен t^beta2_decay",
    },
    eps1: { default: 1e-15, description: "нижний порог оценки второго момента — численная защита знаменателя" },
    eps2: { default: 0.001, description: "нижний порог масштаба параметра: не даёт шагу вырождаться около начала координат" },
    d: { default: 1.0, description: "порог клиппинга: обновление с RMS выше d пропорционально урезается" },
    weightDecay: { default: 0.0, description: "L2-штраф: тянет точку к началу координат, искажая минимизируемую функцию" },
  },
  createInstance(fn, initialX, params) {
    let x: Vec2 = initialX;
    let step = 0;
    let variance: Vec2 = zero2();
    return {
      get x() { return x; },
      params,
      next() {
        const grad = gradient(fn, x[0], x[1]);
        step += 1;
        // для одномерного параметра PyTorch использует нефакторизованную оценку второго момента
        const oneMinusBeta2T = step ** params.beta2Decay;
        const rhoT = Math.min(params.lr, 1 / Math.sqrt(step));
        const alpha = Math.max(params.eps2, norm2(x) / DIM_SQRT) * rhoT;
        const decayed: Vec2 = [x[0] * (1 - params.lr * params.weightDecay), x[1] * (1 - params.lr * params.weightDecay)];
        variance = [
          variance[0] + (grad[0] ** 2 - variance[0]) * oneMinusBeta2T,
          variance[1] + (grad[1] ** 2 - variance[1]) * oneMinusBeta2T,
        ];
        const update: Vec2 = [
          grad[0] / Math.sqrt(Math.max(variance[0], params.eps1 ** 2)),
          grad[1] / Math.sqrt(Math.max(variance[1], params.eps1 ** 2)),
        ];
        const denom = Math.max(1.0, norm2(update) / (DIM_SQRT * params.d));
        x = [decayed[0] - (alpha / denom) * update[0], decayed[1] - (alpha / denom) * update[1]];
        return { x, value: fn(x[0], x[1]) };
      },
      reset() {
        x = initialX;
        step = 0;
        variance = zero2();
      },
    };
  },
};
